File: src/vodoo/aio/documents.py
"""Async Odoo Documents operations for Vodoo."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from vodoo.aio._domain import AsyncDomainNamespace
from vodoo.documents import (
    _LEGACY_FOLDER_FIELDS,
    _MODERN_FOLDER_FIELDS,
    _decode_document_data,
    _DocumentAttrs,
    _folder_domain,
    _normalize_modern_folders,
    _numeric_id,
    _order_folder_tree,
    _prepare_document_values,
    _require_unique_id,
    _safe_document_filename,
    _uses_document_folder_records,
)
from vodoo.exceptions import RecordNotFoundError, VodooError


class AsyncDocumentNamespace(_DocumentAttrs, AsyncDomainNamespace):
    """Async namespace for the ``documents.document`` model."""

    async def _uses_document_folder_records(self) -> bool:
        fields = await self._client.fields_get(
            self._model, fields=["type"], attributes=["selection"]
        )
        return _uses_document_folder_records(fields)

    async def folders(self, limit: int | None = 50, *, tree: bool = False) -> list[dict[str, Any]]:
        """List accessible folders, normalizing parent relationships across Odoo versions."""
        modern = await self._uses_document_folder_records()
        model = self._model if modern else "documents.folder"
        fields = list(_MODERN_FOLDER_FIELDS if modern else _LEGACY_FOLDER_FIELDS)
        records = await self._client.search_read(
            model,
            domain=_folder_domain(modern),
            fields=fields,
            limit=limit,
            order="name, id",
        )
        folders = _normalize_modern_folders(records) if modern else records
        return _order_folder_tree(folders) if tree else folders

    async def resolve_folder(self, folder: int | str) -> int:
        """Resolve a positive folder ID or exact folder name to its record ID."""
        numeric_id = _numeric_id(folder)
        if numeric_id is not None:
            return numeric_id
        return await self._resolve_folder(str(folder), None)

    async def _resolve_folder(self, folder: int | str | None, folder_id: int | None) -> int:
        if (folder is None) == (folder_id is None):
            raise VodooError("Specify exactly one of folder or folder_id")
        if folder_id is not None:
            if folder_id <= 0:
                raise VodooError("folder_id must be a positive integer")
            return folder_id

        assert folder is not None
        numeric_id = _numeric_id(folder)
        if numeric_id is not None:
            return numeric_id

        folder_name = str(folder)
        modern = await self._uses_document_folder_records()
        model = self._model if modern else "documents.folder"
        records = await self._client.search_read(
            model,
            domain=_folder_domain(modern, folder_name),
            fields=["id", "name"],
            limit=2,
        )
        return _require_unique_id(records, model, folder_name)

    async def _resolve_named_record(self, model: str, value: str | int) -> int:
        numeric_id = _numeric_id(value)
        if numeric_id is not None:
            return numeric_id
        text = str(value)
        fields = ["id", "name"]
        if model == "res.users":
            domain: list[Any] = ["|", ("login", "=", text), ("name", "=", text)]
            fields.append("login")
        else:
            domain = [("name", "=", text)]
        records = await self._client.search_read(model, domain=domain, fields=fields, limit=2)
        return _require_unique_id(records, model, text)

    async def upload(
        self,
        file_path: Path | str,
        *,
        folder: int | str | None = None,
        folder_id: int | None = None,
        tags: list[str | int] | None = None,
        owner: str | int | None = None,
        name: str | None = None,
    ) -> int:
        """Upload a file, specifying exactly one of ``folder`` or ``folder_id``.

        Optional tags and owner may be resolved by positive ID or exact name.
        """
        resolved_folder_id = await self._resolve_folder(folder, folder_id)
        tag_ids = [await self._resolve_named_record("documents.tag", tag) for tag in tags or []]
        owner_id = (
            await self._resolve_named_record("res.users", owner) if owner is not None else None
        )
        values = _prepare_document_values(
            file_path,
            folder_id=resolved_folder_id,
            name=name,
            tag_ids=tag_ids,
            owner_id=owner_id,
        )
        return await self._client.create(self._model, values)

    async def download_file(self, document_id: int, output: Path | str | None = None) -> Path:
        """Download a document and return the resolved output path."""
        records = await self._client.read(
            self._model,
            [document_id],
            fields=["name", "type", "file_size", "datas"],
        )
        if not records:
            raise RecordNotFoundError(self._model, document_id)

        document = records[0]
        data = _decode_document_data(document, document_id)
        filename = _safe_document_filename(document.get("name"), document_id)
        output_path = Path(output) if output is not None else Path.cwd() / filename
        if output_path.is_dir():
            output_path /= filename
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_bytes(data)
        return output_path.resolve()

File: packages/typescript/src/namespaces/documents.ts
import { binaryBytes, decodeBase64, encodeBase64 } from "../binary.js";
import type { BinaryInput } from "../binary.js";
import type { OdooClientApi } from "../client-api.js";
import { Cmd } from "../commands.js";
import { RecordNotFoundError, VodooError } from "../errors.js";
import { GeneratedDocumentNamespace } from "../generated/documents.js";
import type { Domain, OdooRecord } from "../types.js";
import type { ListOptions } from "./domain.js";

const LEGACY_FOLDER_FIELDS = ["id", "name", "parent_folder_id"] as const;
const MODERN_FOLDER_FIELDS = ["id", "name", "folder_id"] as const;
const WINDOWS_RESERVED_NAMES = new Set([
  "CON",
  "PRN",
  "AUX",
  "NUL",
  ...Array.from({ length: 9 }, (_, index) => `COM${index + 1}`),
  ...Array.from({ length: 9 }, (_, index) => `LPT${index + 1}`),
]);

export interface DocumentRecord extends OdooRecord {
  id: number;
  name: string;
  create_date?: Date | null;
}

export interface DocumentFolder extends OdooRecord {
  id: number;
  name: string;
  parent_folder_id?: unknown;
  depth?: number;
  path?: string;
}

export interface FolderListOptions {
  limit?: number | null;
  tree?: boolean;
}

export interface DocumentUploadOptions {
  folder?: number | string;
  folderId?: number;
  tags?: readonly (number | string)[];
  owner?: number | string;
  mimetype?: string;
}

export interface DownloadedDocument {
  id: number;
  name: string;
  data: Uint8Array;
  mimetype?: string;
}

function selectionIncludesFolder(fields: Record<string, unknown>): boolean {
  const type = fields.type;
  if (typeof type !== "object" || type === null) return false;
  const selection = (type as Record<string, unknown>).selection;
  if (
    typeof selection === "object" &&
    selection !== null &&
    !Array.isArray(selection)
  ) {
    return Object.hasOwn(selection, "folder");
  }
  return (
    Array.isArray(selection) &&
    selection.some(
      (option) =>
        Array.isArray(option) && option.length > 0 && option[0] === "folder",
    )
  );
}

function relationId(value: unknown): number | null {
  if (typeof value === "number" && Number.isInteger(value)) return value;
  if (Array.isArray(value) && typeof value[0] === "number") return value[0];
  return null;
}

function numericId(value: number | string): number | null {
  const text = typeof value === "string" ? value.trim() : "";
  const parsed =
    typeof value === "number"
      ? value
      : /^[+-]?\d+$/u.test(text)
        ? Number(text)
        : null;
  if (parsed === null) return null;
  if (!Number.isSafeInteger(parsed) || parsed <= 0) {
    throw new VodooError("Record IDs must be positive integers");
  }
  return parsed;
}

function requireUniqueId(
  records: readonly OdooRecord[],
  model: string,
  value: string,
): number {
  if (records.length === 0) {
    throw new VodooError(`No ${model} record found matching '${value}'`);
  }
  if (records.length > 1) {
    throw new VodooError(
      `Multiple ${model} records match '${value}'; use a numeric ID`,
    );
  }
  const recordId = records[0]?.id;
  if (typeof recordId !== "number") {
    throw new VodooError(
      `${model} record matching '${value}' has no numeric ID`,
    );
  }
  return recordId;
}

function folderDomain(modern: boolean, name?: string): Domain {
  const domain: Array<Domain[number]> = modern ? [["type", "=", "folder"]] : [];
  if (name !== undefined) domain.push(["name", "=", name]);
  return domain;
}

export function orderFolderTree(
  folders: readonly DocumentFolder[],
): DocumentFolder[] {
  const byId = new Map(folders.map((folder) => [folder.id, folder]));
  const children = new Map<number | null, DocumentFolder[]>();
  for (const folder of folders) {
    const relation = relationId(folder.parent_folder_id);
    const parentId = relation !== null && byId.has(relation) ? relation : null;
    const siblings = children.get(parentId) ?? [];
    siblings.push(folder);
    children.set(parentId, siblings);
  }
  for (const siblings of children.values()) {
    siblings.sort(
      (left, right) =>
        left.name.localeCompare(right.name, "und", { sensitivity: "base" }) ||
        left.id - right.id,
    );
  }
  const result: DocumentFolder[] = [];
  const visited = new Set<number>();
  const visit = (
    folder: DocumentFolder,
    depth: number,
    parents: readonly string[],
  ): void => {
    if (visited.has(folder.id)) return;
    visited.add(folder.id);
    result.push({
      ...folder,
      depth,
      path: [...parents, folder.name].join(" / "),
    });
    for (const child of children.get(folder.id) ?? []) {
      visit(child, depth + 1, [...parents, folder.name]);
    }
  };
  for (const root of children.get(null) ?? []) visit(root, 0, []);
  for (const folder of [...folders].sort(
    (left, right) =>
      left.name.localeCompare(right.name, "und", { sensitivity: "base" }) ||
      left.id - right.id,
  )) {
    visit(folder, 0, []);
  }
  return result;
}

export function safeDocumentFilename(
  name: unknown,
  documentId: number,
): string {
  const parts = String(name ?? "")
    .replaceAll("\\", "/")
    .split("/");
  let filename = parts.at(-1) ?? "";
  filename = filename
    .replace(/[<>:"/\\|?*\u0000-\u001f]/gu, "_")
    .replace(/[. ]+$/u, "");
  const stem = filename.split(".", 1)[0]?.toLocaleUpperCase("und") ?? "";
  if (WINDOWS_RESERVED_NAMES.has(stem)) filename = `_${filename}`;
  return filename === "" || filename === "." || filename === ".."
    ? `document_${documentId}`
    : filename;
}

function guessMimetype(name: string): string {
  const extension = name.split(".").at(-1)?.toLocaleLowerCase("und");
  const known: Readonly<Record<string, string>> = {
    csv: "text/csv",
    gif: "image/gif",
    html: "text/html",
    jpeg: "image/jpeg",
    jpg: "image/jpeg",
    json: "application/json",
    md: "text/markdown",
    pdf: "application/pdf",
    png: "image/png",
    svg: "image/svg+xml",
    txt: "text/plain",
    webp: "image/webp",
    xml: "application/xml",
    zip: "application/zip",
  };
  return extension === undefined
    ? "application/octet-stream"
    : (known[extension] ?? "application/octet-stream");
}

export class DocumentNamespace extends GeneratedDocumentNamespace {
  private modernFolders: boolean | undefined;

  constructor(client: OdooClientApi) {
    super(client);
  }

  override async list(options: ListOptions = {}): Promise<DocumentRecord[]> {
    return (await super.list(options)) as DocumentRecord[];
  }

  override async get(
    recordId: number,
    fields?: readonly string[],
  ): Promise<DocumentRecord> {
    return (await super.get(recordId, fields)) as DocumentRecord;
  }

  async folders(options: FolderListOptions = {}): Promise<DocumentFolder[]> {
    const modern = await this.usesDocumentFolderRecords();
    const model = modern ? "documents.document" : "documents.folder";
    const records = await this.client.searchRead(model, {
      domain: folderDomain(modern),
      fields: modern ? MODERN_FOLDER_FIELDS : LEGACY_FOLDER_FIELDS,
      limit: options.limit === undefined ? 50 : options.limit,
      order: "name, id",
    });
    const folders = records
      .filter(
        (record): record is OdooRecord & { id: number; name: string } =>
          typeof record.id === "number" && typeof record.name === "string",
      )
      .map<DocumentFolder>((record) =>
        modern
          ? {
              id: record.id,
              name: record.name,
              parent_folder_id: record.folder_id,
            }
          : { ...record },
      );
    return options.tree === true ? orderFolderTree(folders) : folders;
  }

  resolveFolder(folder: number | string): Promise<number> {
    return this.resolveFolderInput({ folder });
  }

  async upload(
    data: BinaryInput,
    name: string,
    options: DocumentUploadOptions,
  ): Promise<number> {
    const folderId = await this.resolveFolderInput(options);
    const tagIds: number[] = [];
    for (const tag of options.tags ?? []) {
      tagIds.push(await this.resolveNamedRecord("documents.tag", tag));
    }
    const ownerId =
      options.owner === undefined
        ? undefined
        : await this.resolveNamedRecord("res.users", options.owner);
    const bytes = await binaryBytes(data);
    const values: Record<string, unknown> = {
      name,
      datas: encodeBase64(bytes),
      mimetype:
        options.mimetype ??
        (typeof Blob !== "undefined" && data instanceof Blob && data.type !== ""
          ? data.type
          : guessMimetype(name)),
      folder_id: folderId,
    };
    if (tagIds.length > 0) values.tag_ids = [Cmd.set(tagIds)];
    if (ownerId !== undefined) values.owner_id = ownerId;
    return this.client.create("documents.document", values);
  }

  /** Worker adaptation of Python download_file: return sanitized name and bytes. */
  async downloadFile(documentId: number): Promise<DownloadedDocument> {
    const records = await this.client.read(
      "documents.document",
      [documentId],
      ["name", "type", "file_size", "datas", "mimetype"],
    );
    const document = records[0];
    if (document === undefined) {
      throw new RecordNotFoundError("documents.document", documentId);
    }
    let data: Uint8Array;
    if (typeof document.datas === "string") {
      data = decodeBase64(document.datas);
    } else if (document.type === "binary" && document.file_size === 0) {
      data = new Uint8Array();
    } else {
      throw new RecordNotFoundError("documents.document", documentId);
    }
    const result: DownloadedDocument = {
      id: documentId,
      name: safeDocumentFilename(document.name, documentId),
      data,
    };
    return typeof document.mimetype === "string"
      ? { ...result, mimetype: document.mimetype }
      : result;
  }

  private async usesDocumentFolderRecords(): Promise<boolean> {
    if (this.modernFolders !== undefined) return this.modernFolders;
    const fields = await this.client.fieldsGet(
      "documents.document",
      ["type"],
      ["selection"],
    );
    this.modernFolders = selectionIncludesFolder(fields);
    return this.modernFolders;
  }

  private async resolveFolderInput(options: {
    folder?: number | string;
    folderId?: number;
  }): Promise<number> {
    if ((options.folder === undefined) === (options.folderId === undefined)) {
      throw new VodooError("Specify exactly one of folder or folderId");
    }
    if (options.folderId !== undefined) {
      const direct = numericId(options.folderId);
      if (direct === null)
        throw new VodooError("folderId must be a positive integer");
      return direct;
    }
    const folder = options.folder;
    if (folder === undefined) throw new VodooError("Specify a folder");
    const direct = numericId(folder);
    if (direct !== null) return direct;
    if (typeof folder !== "string") throw new VodooError("Invalid folder name");
    const modern = await this.usesDocumentFolderRecords();
    const model = modern ? "documents.document" : "documents.folder";
    const records = await this.client.searchRead(model, {
      domain: folderDomain(modern, folder),
      fields: ["id", "name"],
      limit: 2,
    });
    return requireUniqueId(records, model, folder);
  }

  private async resolveNamedRecord(
    model: string,
    value: number | string,
  ): Promise<number> {
    const direct = numericId(value);
    if (direct !== null) return direct;
    if (typeof value !== "string")
      throw new VodooError(`Invalid ${model} name`);
    const domain: Domain =
      model === "res.users"
        ? ["|", ["login", "=", value], ["name", "=", value]]
        : [["name", "=", value]];
    const fields =
      model === "res.users" ? ["id", "name", "login"] : ["id", "name"];
    const records = await this.client.searchRead(model, {
      domain,
      fields,
      limit: 2,
    });
    return requireUniqueId(records, model, value);
  }
}
